web/pgadmin/static/js/TreeView/index.jsx

```diff
@@ -0,0 +1,250 @@
+    if (hasCheckbox) {
+      let selectedChildNodes = [];
+      treeObj.current.selectedNodes.forEach((node) => {
+        selectedChildNodes.push(node.id);
+      });
+      setSelectedCheckBoxNodes(selectedChildNodes);
+    }
+
+    selectionChange?.(treeObj.current.selectedNodes);
+  };
+
+  return (
+    <PgTreeSelectionContext.Provider value={_.isUndefined(selectedCheckBoxNodes) ? []: selectedCheckBoxNodes}>
+      <div className={clsx(classes.tree)}>
+        <TreeView data={treeData} treeRef={treeObj} hasCheckbox={hasCheckbox} eventBusObj={eventBusObj} onSelectionChange={onSelectionChange} {...props}></TreeView>
+      </div>
+    </PgTreeSelectionContext.Provider>
+  );
+}
+
+PgTreeView.propTypes = {
+  data: PropTypes.array,
+  eventBusObj: PropTypes.object,
+  selectionChange: PropTypes.func,
+  hasCheckbox: PropTypes.bool,
+};
+
+function Node({ node, style, dragHandle, tree, hasCheckbox, onNodeSelectionChange}) {
+  const classes = useStyles();
+  const pgTreeSelCtx = React.useContext(PgTreeSelectionContext);
+  const [isSelected, setIsSelected] = React.useState(pgTreeSelCtx.includes(node.id) ? true : false);
+
+
+  const onCheckboxSelection = (e) => {
+    if (hasCheckbox) {
+      setIsSelected(e.currentTarget.checked);
+      if (e.currentTarget.checked) {
+
+        node.selectMulti(node.id);
+        if (!node.isLeaf && node.isOpen) {
+          selectAllChild(node, tree);
+        } else {
+          if (node?.parent) {
+            checkAndSelectParent(node);
+          }
+        }
+      } else {
+        node.deselect(node);
+        if (!node.isLeaf) {
+          deselectAllChild(node);
+        }
+
+        if(node?.parent){
+          delectPrentNode(node.parent);
+        }
+      }
+    }
+
+    onNodeSelectionChange();
+  };
+
+  return (
+    <div style={style} ref={dragHandle} className={clsx(node.isFocused ? classes.focusedNode : '', node.isSelected ? classes.selectedNode : '')} onClick={(e) => {
+      node.focus();
+      e.stopPropagation();
+    }}>
+      <FolderArrow node={node} tree={tree} />
+      {
+        hasCheckbox ? <Checkbox style={{ padding: 0 }} color="primary" className={classes.checkboxStyle} checked={isSelected ? true: false}
+          onChange={onCheckboxSelection}/> :
+          <span className={clsx(node.data.icon, classes.nodeIcon)}></span>
+      }
+      <div className={clsx(node.data.icon, classes.node)}>{node.data.name}</div>
+    </div>
+  );
+}
+
+Node.propTypes = {
+  node: PropTypes.object,
+  style: PropTypes.any,
+  dragHandle: PropTypes.any,
+  tree: PropTypes.object,
+  hasCheckbox: PropTypes.bool,
+  onNodeSelectionChange: PropTypes.func
+};
+
+function FolderArrow({ node, tree }) {
+  const toggleNode = () => {
+    node.isInternal && node.toggle();
+    if (node.isSelected && node.isOpen) {
+      setTimeout(()=>{
+        selectAllChild(node, tree);
+      }, 0);
+
+    }
+  };
+  return (
+    <span onClick={toggleNode} >
+      {node.isInternal ? (
```

Comment:
do not use nested ternary, it will upset sonarqube and difficult to read as well.